Add ListArtifacts

josedonizetti opened the pull request:
Fixes for https://github.com/aquasecurity/trivy-kubernetes/issues/2 and https://github.com/aquasecurity/trivy-kubernetes/issues/1

Example usage (ignoring errors on purpose):
```
ctx := context.Background()
kubeConfig, _ := k8s.GetKubeConfig()
k8sClient, _ := k8s.NewDynamicClient(kubeConfig)

trivyk8s := trivyk8s.New(k8sClient)

//trivy k8s #cluster
artifacts, _ := trivyk8s.ListArtifacts(ctx)
## scan artifacs

// trivy k8s --namespace default
artifacts, _ = trivyk8s.Namespace("default").ListArtifacts(ctx)
### scan artifacts
```

chen-keinan commented:
@josedonizetti printer formatting definitely should be part of `Trivy Project` and not here as it is CLI related

chen-keinan commented on pkg/artifacts/artifacts.go:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	// we don't check found here, if the name is not found it will be an empty string
+	name, _, err := unstructured.NestedString(resource.Object, "metadata", "name")
```
need to check that `len(images) > 0`

josedonizetti replied:
only if we change if we split ListArtifacts in two methods, one for vuln and another for iac. Because for now we have resources passing through this factory method that doesn't have an image, only `metadata: {name: ''}`

chen-keinan replied:
I also added a comment that I think we should split it into two methods

chen-keinan commented on pkg/reportk8s/reportk8s.go:
```diff
@@ -0,0 +1,116 @@
+	"github.com/olekukonko/tablewriter"
+)
+
+type KubernetesReport struct {
```
report should be done on trivy CLI as it not the lib concern

josedonizetti replied:
yes, I agree, my thinking was that for a first version, we have more flexibility having the logic here. But happy to already start it on trivy.

chen-keinan commented on pkg/trivyk8s/trivyk8s.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+// New creates a trivyK8S client
+func New(kubeConfig *rest.Config) (TrivyK8S, error) {
```
I would inject the dynamic client on to this constructor; it could help with testing (using fake client).

josedonizetti replied:
Yes, I'm using the fake client on the tests.

chen-keinan replied:
I assume that you'll have to change the New() function signature to accept client as well

chen-keinan commented on pkg/k8s/k8s.go:
```diff
@@ -0,0 +1,155 @@
+	KindReplicaSet = "ReplicaSet"
+
+	AppsGroup              = "apps"
+	CoreGroup              = "cores"
```
Remove CoreGroup const, if not in use

chen-keinan commented on pkg/trivyk8s/trivyk8s.go:
```diff
@@ -0,0 +1,99 @@
+
+// TrivyK8S interface represents the operations supported by the library
+type TrivyK8S interface {
+	ListArtifacts(context.Context, string) ([]*artifacts.Artifact, error)
```
We might need another method for Iac resources. 
Iac scanner resources -  scan more resources than `workload` , such configmaps ,resourcequota ,randlimit and more.
Vulnerability scanner -  resources scan only workload resources
Basically, we need to fetch all supported resources for the namespace.
- for iac scanner  returns all resources
- for vulnerability scanner  return only workload resources

josedonizetti replied:
I started with two separated methods, endup changing to one that returns everything because we were not sure if we wanted `trivy k8s` to support flags for specific scans (only iac, or only vulns). If `trivy k8s` will always do both scans at the same time and only have flags for `namespace`, or `cluster`, I think having only want method express better the intention of everything is a k8s scannable artifact, and trivy will run different scans on it.

chen-keinan replied:
Regardless of the flags, resources are different for iac and vulnerability scanners. No reason for  vulnerability scanner to accept iac resources

josedonizetti replied:
vulnerability scanner is not accepting it, it only scan images and resources do have images, and others don't.

chen-keinan commented on pkg/artifacts/artifacts.go:
```diff
@@ -0,0 +1,77 @@
+		return nil, err
+	}
+
+	return &Artifact{
```
for Iac we need to return the actual object in json format

chen-keinan commented on pkg/artifacts/artifacts.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+// FromResource is a factory method to create an Artifact from an unstructured.Unstructured
+func FromResource(resource unstructured.Unstructured) (*Artifact, error) {
```
I assume this method is relevant for vulnerability scanners only

josedonizetti replied:
if we change the code to have `iac` and `vuln` as two separate things yes, but for now it will return an artifact with the proper name from metadata, even when it doesn't have an image.

josedonizetti commented:
@chen-keinan Can you take another look?  I have added a new method for `namespace` filtering to avoid passing empty strings. Also, refactored part of the k8s code. One thing though, I didn't split the `ListArtifacts` in two methods. 

I'm not seen the use case to have a different method to list workloads. From trivy perspective it now supports a full cluster scan or a namespace scan. For both cases it need all resources in the cluster or in the namespace, and depending on the resource it calls vulnerabilities scan or iac scan. Why would we split in two methods if trivy doesn't need it? 

Also removed the method `WriteToFile` for now, because I think it is more of trivy responsibility, because trivy is the one needing a file in the FS to be able to scan the resource.

Let me know what you think please.

josedonizetti commented on pkg/k8s/k8s.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+// GetCurrentContext returns local kubernetes current-context
+func GetCurrentContext() (string, error) {
```
It is use on trivy for the cluster name on the report -> https://github.com/aquasecurity/trivy/pull/2065/files#diff-266c01e5e4ea90c70796db35de861161d5319960f2973576b506e4c7423b2614R83

chen-keinan replied:
@josedonizetti if it use in Trivy, then, maybe it should be interfaced
